File: snakes-and-ladders-1/solution.go
package main

import "fmt"

type SnakesLadders struct {
	player1Pos    int
	player2Pos    int
	currentPlayer int
	gameWon       bool
	snakes        map[int]int
	ladders       map[int]int
}

func NewSnakesLadders() *SnakesLadders {
	return &SnakesLadders{
		player1Pos:    0,
		player2Pos:    0,
		currentPlayer: 1,
		gameWon:       false,
		snakes: map[int]int{
			17: 7, 54: 34, 62: 19, 88: 24, 93: 73, 95: 75, 98: 79, 99: 40,
		},
		ladders: map[int]int{
			2: 38, 7: 14, 15: 26, 21: 42, 28: 84, 51: 67, 72: 91, 78: 98, 87: 94,
		},
	}
}

func (sl *SnakesLadders) Play(die1, die2 int) string {
	if sl.gameWon {
		return "Game over!"
	}

	steps := die1 + die2
	var pos int

	if sl.currentPlayer == 1 {
		pos = sl.player1Pos
	} else {
		pos = sl.player2Pos
	}

	movedPlayer := sl.currentPlayer

	pos += steps

	// bounce off 100
	if pos > 100 {
		pos = 100 - (pos - 100)
	}

	// apply snake or ladder
	if snake, ok := sl.snakes[pos]; ok {
		pos = snake
	} else if ladder, ok := sl.ladders[pos]; ok {
		pos = ladder
	}

	if sl.currentPlayer == 1 {
		sl.player1Pos = pos
	} else {
		sl.player2Pos = pos
	}

	if pos == 100 {
		sl.gameWon = true
		return fmt.Sprintf("Player %d Wins!", movedPlayer)
	}

	if die1 != die2 {
		sl.currentPlayer = 3 - sl.currentPlayer
	}

	return fmt.Sprintf("Player %d is on square %d", movedPlayer, pos)
}

File: snakes-and-ladders-1/solution.js
class SnakesLadders {
  constructor() {
    this.positions = [0, 0];
    this.current = 0;
    this.winner = null;
    this.ladders = {
      1: 38, 4: 14, 9: 31, 21: 42, 28: 84,
      36: 44, 51: 67, 71: 91, 80: 100
    };
    this.snakes = {
      16: 6, 47: 26, 49: 11, 56: 53, 62: 19,
      64: 60, 87: 24, 93: 73, 95: 75, 98: 78
    };
  }
  
  play(die1, die2) {
    if (this.winner !== null) return "Game over!";
    
    const player = this.current;
    let pos = this.positions[player] + die1 + die2;
    if (pos > 100) pos = 200 - pos;
    if (this.ladders[pos] !== undefined) pos = this.ladders[pos];
    else if (this.snakes[pos] !== undefined) pos = this.snakes[pos];
    
    this.positions[player] = pos;
    
    if (pos === 100) {
      this.winner = player + 1;
      return `Player ${player + 1} Wins!`;
    }
    
    const message = `Player ${player + 1} is on square ${pos}`;
    
    if (die1 !== die2) this.current = 1 - this.current;
    
    return message;
  }
}